File: Remoter-Mac/Sources/RemoterAgent/Protocol.swift
import Foundation

// Binary frame type prefix (first byte of binary WebSocket messages from server)
enum FrameType: UInt8 {
    case videoFrame = 0x01
    case fileChunk  = 0x02
    case audioFrame = 0x03   // [0x03][ADTS-framed AAC packet]
}

// Incoming JSON message types from client
enum ClientMessage {
    case auth(pin: String)
    case mouseMove(x: Double, y: Double, dragging: String?)
    case mouseButton(button: String, down: Bool, x: Double, y: Double)
    case mouseDoubleClick(button: String, x: Double, y: Double)
    case mouseScroll(dx: Int, dy: Int)
    case key(code: String, down: Bool, modifiers: [String])
    case textInput(String)   // IME-composed text, injected as a unicode string
    case clipboardSet(text: String)
    case fileStart(id: String, name: String, size: Int64)
    case fileEnd(id: String)
    case fpsSet(fps: Int, auto: Bool)
    case bitrateSet(bitrate: Int, auto: Bool)
    case resolutionSet(tier: String)
    case setAudioEnabled(Bool)
    case displaySet(id: UInt32)
    case ping
    case requestKeyframe
    case webrtcOffer(sdp: String)
    case webrtcICE(json: String)
    case clientStats(fps: Double, rttMs: Int)
    case setCodec(codec: String)
    case cryptoHello(pubkey: String)
    case listDir(path: String)
    case requestFile(path: String)
    case setMuted(muted: Bool)
    case ctrlAltDel
    case setClipboardSync(enabled: Bool)
    case setInputEnabled(enabled: Bool)
    case clipboardSetImage(data: String)
    case lockScreen
    case logout
    case restart
    case authCredentials(username: String, password: String)
    case authToken(token: String)
    case unknown

    static func parse(_ json: [String: Any]) -> ClientMessage {
        guard let type = json["type"] as? String else { return .unknown }
        switch type {
        case "auth":
            return .auth(pin: json["pin"] as? String ?? "")
        case "mouse_move":
            return .mouseMove(
                x: json["x"] as? Double ?? 0,
                y: json["y"] as? Double ?? 0,
                dragging: json["dragging"] as? String
            )
        case "mouse_button":
            return .mouseButton(
                button: json["button"] as? String ?? "left",
                down: json["down"] as? Bool ?? false,
                x: json["x"] as? Double ?? 0,
                y: json["y"] as? Double ?? 0
            )
        case "mouse_double_click":
            return .mouseDoubleClick(
                button: json["button"] as? String ?? "left",
                x: json["x"] as? Double ?? 0,
                y: json["y"] as? Double ?? 0
            )
        case "mouse_scroll":
            return .mouseScroll(
                dx: json["dx"] as? Int ?? 0,
                dy: json["dy"] as? Int ?? 0
            )
        case "key":
            return .key(
                code: json["code"] as? String ?? "",
                down: json["down"] as? Bool ?? false,
                modifiers: json["modifiers"] as? [String] ?? []
            )
        case "text_input":
            return .textInput(json["text"] as? String ?? "")
        case "clipboard_set":
            return .clipboardSet(text: json["text"] as? String ?? "")
        case "file_start":
            return .fileStart(
                id: json["id"] as? String ?? "",
                name: json["name"] as? String ?? "",
                size: Int64(json["size"] as? Int ?? 0)
            )
        case "file_end":
            return .fileEnd(id: json["id"] as? String ?? "")
        case "fps":
            return .fpsSet(
                fps: json["fps"] as? Int ?? 30,
                auto: json["auto"] as? Bool ?? false
            )
        case "bitrate":
            return .bitrateSet(
                bitrate: json["bitrate"] as? Int ?? 2_000_000,
                auto: json["auto"] as? Bool ?? false
            )
        case "resolution":
            return .resolutionSet(tier: json["tier"] as? String ?? "1080")
        case "set_audio":
            return .setAudioEnabled(json["enabled"] as? Bool ?? false)
        case "display":
            return .displaySet(id: UInt32(json["id"] as? Int ?? 0))
        case "ping":
            return .ping
        case "request_keyframe":
            return .requestKeyframe
        case "webrtc_offer":
            return .webrtcOffer(sdp: json["sdp"] as? String ?? "")
        case "webrtc_ice":
            return .webrtcICE(json: json["candidate"] as? String ?? "")
        case "client_stats":
            return .clientStats(
                fps:   json["fps"]    as? Double ?? 0,
                rttMs: json["rtt_ms"] as? Int    ?? 0
            )
        case "set_codec":
            return .setCodec(codec: json["codec"] as? String ?? "h264")
        case "crypto_hello":
            return .cryptoHello(pubkey: json["pubkey"] as? String ?? "")
        case "list_dir":
            return .listDir(path: json["path"] as? String ?? "~")
        case "request_file":
            return .requestFile(path: json["path"] as? String ?? "")
        case "set_muted":
            return .setMuted(muted: json["muted"] as? Bool ?? false)
        case "ctrl_alt_del":
            return .ctrlAltDel
        case "set_clipboard_sync":
            return .setClipboardSync(enabled: json["enabled"] as? Bool ?? true)
        case "set_input_enabled":
            return .setInputEnabled(enabled: json["enabled"] as? Bool ?? true)
        case "clipboard_set_image":
            return .clipboardSetImage(data: json["data"] as? String ?? "")
        case "lock_screen":
            return .lockScreen
        case "logout":
            return .logout
        case "restart":
            return .restart
        case "auth_credentials":
            return .authCredentials(
                username: json["username"] as? String ?? "",
                password: json["password"] as? String ?? "")
        case "auth_token":
            return .authToken(token: json["token"] as? String ?? "")
        default:
            return .unknown
        }
    }
}

func buildVideoFramePacket(data: Data, frameId: UInt32, ptsMs: UInt32, isKeyframe: Bool) -> Data {
    var packet = Data(capacity: 10 + data.count)
    packet.append(FrameType.videoFrame.rawValue)
    var fid = frameId.bigEndian
    packet.append(Data(bytes: &fid, count: 4))
    var pts = ptsMs.bigEndian
    packet.append(Data(bytes: &pts, count: 4))
    packet.append(isKeyframe ? 0x01 : 0x00)
    packet.append(data)
    return packet
}

File: Remoter-Win/Protocol.cs
using System.Text.Json;

namespace RemoterWin;

// Mirrors the Mac agent's ClientMessage enum.
// Discriminated union via C# record hierarchy.
public abstract record ClientMsg
{
    public record Auth(string Pin) : ClientMsg;
    public record MouseMove(double X, double Y) : ClientMsg;
    public record MouseButton(string Button, bool Down, double X, double Y) : ClientMsg;
    public record MouseScroll(int Dx, int Dy) : ClientMsg;
    public record KeyEvent(string Code, bool Down, string[] Mods) : ClientMsg;
    public record ClipboardSet(string Text) : ClientMsg;
    public record FileStart(string Id, string Name, long Size) : ClientMsg;
    public record FileEnd(string Id) : ClientMsg;
    public record QualitySet(int Fps, int Bitrate) : ClientMsg;   // legacy clients (≤ v1.0.204)
    public record FpsSet(int Fps, bool Auto) : ClientMsg;
    public record BitrateSet(int Bitrate, bool Auto) : ClientMsg;
    public record ResolutionSet(string Tier) : ClientMsg;
    public record Ping : ClientMsg;
    public record RequestKeyframe : ClientMsg;
    public record WebRtcOffer(string Sdp) : ClientMsg;
    public record WebRtcIce(string Candidate) : ClientMsg;
    public record ClientStats(double Fps, int RttMs) : ClientMsg;
    public record SetCodec(string Codec) : ClientMsg;
    public record CryptoHello(string Pubkey) : ClientMsg;
    public record ListDir(string Path) : ClientMsg;
    public record RequestFile(string Path) : ClientMsg;
    public record SetMuted(bool Muted) : ClientMsg;
    public record CtrlAltDel : ClientMsg;
    public record SetInputEnabled(bool Enabled) : ClientMsg;
    public record ClipboardSetImage(string Data) : ClientMsg;
    public record LockScreen : ClientMsg;
    public record Logout : ClientMsg;
    public record Restart : ClientMsg;
    public record AuthCredentials(string Username, string Password) : ClientMsg;
    public record AuthToken(string Token) : ClientMsg;
    public record MouseDoubleClick(string Button, double X, double Y) : ClientMsg;
    public record Unknown : ClientMsg;

    public static ClientMsg Parse(JsonElement e)
    {
        var type = e.Str("type");
        return type switch
        {
            "auth"              => new Auth(e.Str("pin")),
            "mouse_move"        => new MouseMove(e.Dbl("x"), e.Dbl("y")),
            "mouse_button"      => new MouseButton(e.Str("button"), e.Bool("down"), e.Dbl("x"), e.Dbl("y")),
            "mouse_scroll"      => new MouseScroll(e.Int("dx"), e.Int("dy")),
            "key"               => new KeyEvent(e.Str("code"), e.Bool("down"), e.StrArr("modifiers")),
            "clipboard_set"     => new ClipboardSet(e.Str("text")),
            "file_start"        => new FileStart(e.Str("id"), e.Str("name"), e.Long("size")),
            "file_end"          => new FileEnd(e.Str("id")),
            "quality"           => new QualitySet(e.Int("fps"), e.Int("bitrate")),
            "fps"               => new FpsSet(e.Int("fps"), e.Bool("auto")),
            "bitrate"           => new BitrateSet(e.Int("bitrate"), e.Bool("auto")),
            "resolution"        => new ResolutionSet(e.Str("tier")),
            "ping"              => new Ping(),
            "request_keyframe"  => new RequestKeyframe(),
            "webrtc_offer"      => new WebRtcOffer(e.Str("sdp")),
            "webrtc_ice"        => new WebRtcIce(e.Str("candidate")),
            "client_stats"      => new ClientStats(e.Dbl("fps"), e.Int("rtt_ms")),
            "set_codec"         => new SetCodec(e.Str("codec")),
            "crypto_hello"      => new CryptoHello(e.Str("pubkey")),
            "list_dir"          => new ListDir(e.Str("path")),
            "request_file"      => new RequestFile(e.Str("path")),
            "set_muted"         => new SetMuted(e.Bool("muted")),
            "ctrl_alt_del"      => new CtrlAltDel(),
            "set_input_enabled"       => new SetInputEnabled(e.Bool("enabled")),
            "clipboard_set_image"     => new ClipboardSetImage(e.Str("data")),
            "lock_screen"             => new LockScreen(),
            "logout"                  => new Logout(),
            "restart"                 => new Restart(),
            "auth_credentials"        => new AuthCredentials(e.Str("username"), e.Str("password")),
            "auth_token"              => new AuthToken(e.Str("token")),
            "mouse_double_click"      => new MouseDoubleClick(e.Str("button"), e.Dbl("x"), e.Dbl("y")),
            _                         => new Unknown(),
        };
    }
}

// Helpers to read optional JSON properties with defaults
static class JsonExt
{
    public static string  Str (this JsonElement e, string k) =>
        e.TryGetProperty(k, out var v) ? v.GetString() ?? "" : "";
    public static bool    Bool(this JsonElement e, string k) =>
        e.TryGetProperty(k, out var v) && v.GetBoolean();
    public static double  Dbl (this JsonElement e, string k) =>
        e.TryGetProperty(k, out var v) ? v.GetDouble() : 0;
    public static int     Int (this JsonElement e, string k) =>
        e.TryGetProperty(k, out var v) ? v.GetInt32() : 0;
    public static long    Long(this JsonElement e, string k) =>
        e.TryGetProperty(k, out var v) ? v.GetInt64() : 0;
    public static string[] StrArr(this JsonElement e, string k)
    {
        if (!e.TryGetProperty(k, out var v) || v.ValueKind != JsonValueKind.Array) return [];
        return [.. v.EnumerateArray().Select(x => x.GetString() ?? "")];
    }
}

// Binary frame header (same as Mac agent)
static class FrameType
{
    public const byte VideoFrame = 0x01;
    public const byte FileChunk  = 0x02;
    public const byte Encrypted  = 0xE0;
}

static class FrameBuilder
{
    public static byte[] VideoFramePacket(byte[] jpeg, uint frameId, uint ptsMs, bool keyframe)
    {
        var pkt = new byte[10 + jpeg.Length];
        pkt[0] = FrameType.VideoFrame;
        WriteUInt32BE(pkt, 1, frameId);
        WriteUInt32BE(pkt, 5, ptsMs);
        pkt[9] = keyframe ? (byte)1 : (byte)0;
        Buffer.BlockCopy(jpeg, 0, pkt, 10, jpeg.Length);
        return pkt;
    }

    private static void WriteUInt32BE(byte[] buf, int offset, uint value)
    {
        buf[offset]     = (byte)(value >> 24);
        buf[offset + 1] = (byte)(value >> 16);
        buf[offset + 2] = (byte)(value >> 8);
        buf[offset + 3] = (byte)(value);
    }
}
